chainer/functions/evaluation/classification_summary.py

```diff
@@ -0,0 +1,124 @@
+from __future__ import division
+
+import numpy
+import six
+
+import chainer
+from chainer import cuda
+from chainer import function
+from chainer.utils import type_check
+
+
+def _f1_score(precision, recall, beta):
+    beta_square = beta * beta
+    return ((1 + beta_square) * precision * recall /
+            (beta_square * precision + recall)).astype(precision.dtype)
+
+
+class ClassificationSummary(function.Function):
+
+    ignore_label = -1
+
+    def __init__(self, beta, label_num):
+        self.beta = beta
+        self.label_num = label_num
+
+    def check_type_forward(self, in_types):
+        type_check.expect(in_types.size() == 2)
+        x_type, t_type = in_types
+
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            t_type.dtype == numpy.int32
+        )
+
+        t_ndim = t_type.ndim.eval()
+        type_check.expect(
+            x_type.ndim >= t_type.ndim,
+            x_type.shape[0] == t_type.shape[0],
+            x_type.shape[2: t_ndim + 1] == t_type.shape[1:]
+        )
+        for i in six.moves.range(t_ndim + 1, x_type.ndim.eval()):
+            type_check.expect(x_type.shape[i] == 1)
+
+    def forward(self, inputs):
```

Comment:
As you mentioned, the current implementation is insufficient when the number of labels is relatively large.
How about using ``xp.bincount()`` instead of ``for`` statement, and replace it with the following?
```python
    def forward(self, inputs):
        xp = cuda.get_array_module(*inputs)
        y, t = inputs

        if self.label_num is None:
            label_num = xp.maximum(t) + 1
        else:
            label_num = self.label_num
            if chainer.is_debug():
                assert (t < label_num).all()

        mask = (t == self.ignore_label).ravel()
        pred = xp.where(mask, label_num, y.argmax(axis=1).ravel())
        true = xp.where(mask, label_num, t.ravel())
        support = xp.bincount(true, minlength=label_num + 1)[:label_num]
        relevant = xp.bincount(pred, minlength=label_num + 1)[:label_num]
        tp_mask = xp.where(pred == true, true, label_num)
        tp = xp.bincount(tp_mask, minlength=label_num + 1)[:label_num]

        precision = tp / relevant
        recall = tp / support

        f1 = _f1_score(precision, recall, self.beta)

        return precision, recall, f1, support
```
I confirmed that the above code passes the test.